model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsChangelogBasedTransaction.java

```diff
@@ -279,4 +282,95 @@ protected void rollbackImpl() {
 
     }
 
+    /**
+     * Imports a session from an external source into the {@link Cache}.
+     * <p>
+     * If a session already exists in the cache, this method does not insert the {@code session}. The invoker should use
+     * the session returned by this method invocation. When the session is successfully imported, this method returns
+     * null and the {@code session} can be used by the transaction.
+     * <p>
+     * This transaction will keep track of further changes in the session.
+     *
+     * @param realmModel The {@link RealmModel} where the session belong to.
+     * @param key        The cache's key.
+     * @param session    The session to import.
+     * @param lifespan   How long the session stays cached until it is expired and removed.
+     * @param maxIdle    How long the session can be idle (without reading or writing) before being removed.
+     * @param offline    {@code true} if it is an offline session.
+     * @return The existing cached session. If it returns {@code null}, it means the {@code session} used in the
+     * parameters was cached.
+     */
+    public SessionEntityWrapper<V> importSession(RealmModel realmModel, K key, SessionEntityWrapper<V> session, boolean offline, long lifespan, long maxIdle) {
+        var updates = getUpdates(offline);
+        var updatesList = getUpdates(offline).get(key);
+        if (updatesList != null) {
+            // exists in transaction, avoid import operation
+            return updatesList.getEntityWrapper();
+        }
+        SessionEntityWrapper<V> existing = null;
+        try {
+            if (getCache(offline) != null) {
+                existing = getCache(offline).putIfAbsent(key, session, lifespan, TimeUnit.MILLISECONDS, maxIdle, TimeUnit.MILLISECONDS);
+            }
+        } catch (RuntimeException exception) {
+            // debug level, it the import fail the transaction can continue with the data from the database.
```

Comment:
`// If the import fails, the transaction can continue with the data from the database.`